File: tools/sovereign-chatgpt-mcp/deterministic_contract.py
from __future__ import annotations

from copy import deepcopy
import hashlib
import json
import re
import unicodedata
from typing import Any, Mapping


KAPPA_SCALE = 1_000_000
MAX_CANONICAL_BYTES = 250_000
MAX_REPLAY_ACTIONS = 100
_STATE_METADATA_FIELDS = frozenset({
    "stateHash", "actionHash", "chainHash", "previousChainHash",
})


def _normalize_text(value: str, *, path: str) -> str:
    normalized = unicodedata.normalize("NFC", value)
    if any(0xD800 <= ord(character) <= 0xDFFF for character in normalized):
        raise ValueError(f"{path} contains an unpaired Unicode surrogate")
    return normalized


def _utf16_sort_key(value: str) -> bytes:
    return value.encode("utf-16-be")


def canonical_decimal_to_units(value: str, *, signed: bool = True) -> int:
    raw = _normalize_text(str(value or "").strip(), path="value")
    match = re.fullmatch(r"(-?)(0|[1-9][0-9]*)(?:\.([0-9]+))?", raw)
    if match is None:
        raise ValueError("value must be canonical decimal text")
    negative = match.group(1) == "-"
    if negative and not signed:
        raise ValueError("negative values are not allowed")
    whole = int(match.group(2))
    fraction = match.group(3) or ""
    fractional_units = int((fraction[:6].ljust(6, "0") or "0"))
    units = whole * KAPPA_SCALE + fractional_units
    return -units if negative and units else units


def trunc_div_toward_zero(numerator: int, denominator: int) -> int:
    if denominator == 0:
        raise ZeroDivisionError("division by zero")
    quotient = abs(numerator) // abs(denominator)
    return -quotient if (numerator < 0) ^ (denominator < 0) else quotient


def multiply_fixed(left: int, right: int) -> int:
    return trunc_div_toward_zero(left * right, KAPPA_SCALE)


def divide_fixed(left: int, right: int) -> int:
    return trunc_div_toward_zero(left * KAPPA_SCALE, right)


def _canonical_value(value: Any, *, path: str = "$") -> Any:
    if value is None or isinstance(value, bool):
        return value
    if isinstance(value, int) and not isinstance(value, bool):
        return value
    if isinstance(value, float):
        raise TypeError(
            f"{path} contains a float; use scaled integers or canonical decimal text"
        )
    if isinstance(value, str):
        return _normalize_text(value, path=path)
    if isinstance(value, Mapping):
        entries: dict[str, Any] = {}
        for key, item in value.items():
            if not isinstance(key, str):
                raise TypeError(f"{path} contains a non-string mapping key")
            normalized_key = _normalize_text(key, path=f"{path}.<key>")
            if normalized_key in entries:
                raise ValueError(
                    f"{path} contains duplicate keys after Unicode normalization"
                )
            entries[normalized_key] = item
        normalized: dict[str, Any] = {}
        for normalized_key in sorted(entries, key=_utf16_sort_key):
            normalized[normalized_key] = _canonical_value(
                entries[normalized_key], path=f"{path}.{normalized_key}"
            )
        return normalized
    if isinstance(value, (list, tuple)):
        return [
            _canonical_value(item, path=f"{path}[{index}]")
            for index, item in enumerate(value)
        ]
    raise TypeError(f"{path} contains unsupported type {type(value).__name__}")


def canonical_bytes(value: Any) -> bytes:
    normalized = _canonical_value(value)
    payload = json.dumps(
        normalized,
        ensure_ascii=False,
        sort_keys=False,
        separators=(",", ":"),
        allow_nan=False,
    ).encode("utf-8")
    if len(payload) > MAX_CANONICAL_BYTES:
        raise ValueError("canonical payload exceeds the bounded size limit")
    return payload


def canonical_sha256(value: Any) -> str:
    return hashlib.sha256(canonical_bytes(value)).hexdigest()


def state_sha256(value: Mapping[str, Any]) -> str:
    normalized = _canonical_value(value)
    if not isinstance(normalized, dict):
        raise TypeError("state must be a mapping")
    payload = {
        key: item
        for key, item in normalized.items()
        if key not in _STATE_METADATA_FIELDS
    }
    return canonical_sha256(payload)


def normalize_transition_table(value: Any) -> dict[str, dict[str, str]]:
    if not isinstance(value, Mapping) or not value:
        raise ValueError("transition_table must be a non-empty mapping")
    output: dict[str, dict[str, str]] = {}
    for raw_state, raw_actions in value.items():
        if not isinstance(raw_state, str) or not raw_state.strip():
            raise ValueError("transition state names must be non-empty strings")
        if not isinstance(raw_actions, Mapping) or not raw_actions:
            raise ValueError("each transition state must contain action mappings")
        state = _normalize_text(raw_state.strip(), path="transition state")
        if state in output:
            raise ValueError("duplicate transition state after normalization")
        actions: dict[str, str] = {}
        for raw_action, raw_target in raw_actions.items():
            if not isinstance(raw_action, str) or not raw_action.strip():
                raise ValueError("transition action names must be non-empty strings")
            if not isinstance(raw_target, str) or not raw_target.strip():
                raise ValueError("transition targets must be non-empty strings")
            action = _normalize_text(raw_action.strip(), path="transition action")
            target = _normalize_text(raw_target.strip(), path="transition target")
            if action in actions:
                raise ValueError("duplicate transition action after normalization")
            actions[action] = target
        output[state] = {
            key: actions[key]
            for key in sorted(actions, key=_utf16_sort_key)
        }
    return {
        key: output[key]
        for key in sorted(output, key=_utf16_sort_key)
    }


def transition_preview(
    current_state: Mapping[str, Any],
    action: Mapping[str, Any],
    transition_table: Mapping[str, Any],
    *,
    expected_version: int | None = None,
    expected_state_hash: str = "",
    engine_version: str = "are-v1",
) -> dict[str, Any]:
    state = _canonical_value(current_state)
    selected_action = _canonical_value(action)
    table = normalize_transition_table(transition_table)
    if not isinstance(state, dict) or not isinstance(selected_action, dict):
        raise TypeError("current_state and action must be mappings")

    current_status = state.get("status")
    action_type = selected_action.get("type")
    if not isinstance(current_status, str) or not current_status:
        raise ValueError("current_state.status is required")
    if not isinstance(action_type, str) or not action_type:
        raise ValueError("action.type is required")

    current_version = state.get("version", 0)
    if (
        not isinstance(current_version, int)
        or isinstance(current_version, bool)
        or current_version < 0
    ):
        raise ValueError("current_state.version must be a non-negative integer")

    current_hash = state_sha256(state)
    if expected_version is not None and current_version != expected_version:
        return {
            "ok": False,
            "allowed": False,
            "status": "VERSION_CONFLICT",
            "currentVersion": current_version,
            "expectedVersion": expected_version,
            "currentStateHash": current_hash,
            "mutationPerformed": False,
        }
    if expected_state_hash and current_hash != expected_state_hash:
        return {
            "ok": False,
            "allowed": False,
            "status": "STATE_HASH_CONFLICT",
            "currentStateHash": current_hash,
            "expectedStateHash": expected_state_hash,
            "mutationPerformed": False,
        }

    target = table.get(current_status, {}).get(action_type)
    if target is None:
        return {
            "ok": False,
            "allowed": False,
            "status": "TRANSITION_NOT_ALLOWED",
            "currentStatus": current_status,
            "actionType": action_type,
            "allowedActions": sorted(table.get(current_status, {})),
            "currentStateHash": current_hash,
            "mutationPerformed": False,
        }

    requested_target = selected_action.get("targetStatus")
    if requested_target is not None and requested_target != target:
        return {
            "ok": False,
            "allowed": False,
            "status": "TARGET_STATUS_MISMATCH",
            "currentStatus": current_status,
            "actionType": action_type,
            "contractTargetStatus": target,
            "requestedTargetStatus": requested_target,
            "currentStateHash": current_hash,
            "mutationPerformed": False,
        }

    patch = selected_action.get("patch", {})
    if not isinstance(patch, dict):
        raise ValueError("action.patch must be a mapping when supplied")
    protected = {
        "status",
        "version",
        "stateHash",
        "actionHash",
        "chainHash",
        "previousChainHash",
    }
    forbidden = sorted(protected.intersection(patch))
    if forbidden:
        raise ValueError(
            f"action.patch contains protected fields: {', '.join(forbidden)}"
        )

    next_state = deepcopy(state)
    for metadata_field in _STATE_METADATA_FIELDS:
        next_state.pop(metadata_field, None)
    for key in sorted(patch):
        next_state[key] = patch[key]
    next_state["status"] = target
    next_state["version"] = current_version + 1

    action_hash = canonical_sha256(selected_action)
    state_hash = state_sha256(next_state)
    previous_chain_hash = str(state.get("chainHash") or "")
    chain_material = {
        "schemaVersion": "sovereign.are-chain.v1",
        "engineVersion": unicodedata.normalize(
            "NFC", engine_version.strip() or "are-v1"
        ),
        "sequence": next_state["version"],
        "previousChainHash": previous_chain_hash,
        "actionHash": action_hash,
        "stateHash": state_hash,
    }
    chain_hash = canonical_sha256(chain_material)

    return {
        "ok": True,
        "allowed": True,
        "status": "TRANSITION_VALIDATED",
        "currentStatus": current_status,
        "nextStatus": target,
        "currentVersion": current_version,
        "nextVersion": next_state["version"],
        "currentStateHash": current_hash,
        "actionHash": action_hash,
        "nextStateHash": state_hash,
        "previousChainHash": previous_chain_hash,
        "chainHash": chain_hash,
        "stateHashContract": "canonical-state-without-chain-metadata",
        "nextState": next_state,
        "mutationPerformed": False,
        "truthNotice": (
            "Pure transition preview only; persistence, effects and runtime success "
            "are not claimed."
        ),
    }


def replay_verify(
    initial_state: Mapping[str, Any],
    actions: list[dict[str, Any]],
    transition_table: Mapping[str, Any],
    *,
    expected_final_state_hash: str = "",
    engine_version: str = "are-v1",
) -> dict[str, Any]:
    if not isinstance(actions, list):
        raise ValueError("actions must be a list")
    if len(actions) > MAX_REPLAY_ACTIONS:
        raise ValueError(
            f"actions exceed the bounded limit of {MAX_REPLAY_ACTIONS}"
        )
    state = _canonical_value(initial_state)
    if not isinstance(state, dict):
        raise ValueError("initial_state must be a mapping")

    steps: list[dict[str, Any]] = []
    for index, action in enumerate(actions, 1):
        result = transition_preview(
            state,
            action,
            transition_table,
            expected_version=state.get("version", 0),
            expected_state_hash=state_sha256(state),
            engine_version=engine_version,
        )
        steps.append(
            {
                "index": index,
                "allowed": bool(result.get("allowed")),
                "status": result.get("status"),
                "actionHash": result.get("actionHash"),
                "stateHash": result.get("nextStateHash"),
                "chainHash": result.get("chainHash"),
            }
        )
        if not result.get("allowed"):
            return {
                "ok": False,
                "status": "REPLAY_BLOCKED",
                "failedStep": index,
                "steps": steps,
                "finalState": state,
                "finalStateHash": state_sha256(state),
                "finalChainHash": state.get("chainHash") or None,
                "stateHashContract": "canonical-state-without-chain-metadata",
                "mutationPerformed": False,
                "crossRuntimeParityProven": False,
            }
        state = result["nextState"]
        state["chainHash"] = result["chainHash"]

    final_hash = state_sha256(state)
    matches = not expected_final_state_hash or final_hash == expected_final_state_hash
    return {
        "ok": matches,
        "status": "REPLAY_VERIFIED" if matches else "FINAL_STATE_HASH_MISMATCH",
        "steps": steps,
        "finalState": state,
        "finalStateHash": final_hash,
        "finalChainHash": state.get("chainHash") or None,
        "stateHashContract": "canonical-state-without-chain-metadata",
        "expectedFinalStateHash": expected_final_state_hash or None,
        "mutationPerformed": False,
        "crossRuntimeParityProven": False,
        "truthNotice": (
            "Python reference replay only. TypeScript/Python bit parity requires "
            "an independent TypeScript run over the same canonical vectors."
        ),
    }

File: tools/sovereign-chatgpt-mcp/deterministic_contract.ts
import { createHash } from 'node:crypto';

export const KAPPA_SCALE = 1_000_000n;
export const MAX_CANONICAL_BYTES = 250_000;
export const MAX_REPLAY_ACTIONS = 100;

const STATE_METADATA_FIELDS = new Set([
  'stateHash',
  'actionHash',
  'chainHash',
  'previousChainHash',
]);

type CanonicalPrimitive = null | boolean | bigint | string;
export type CanonicalValue =
  | CanonicalPrimitive
  | CanonicalValue[]
  | { [key: string]: CanonicalValue };

export type CanonicalObject = { [key: string]: CanonicalValue };
export type TransitionTable = { [state: string]: { [action: string]: string } };

function assertNoUnpairedSurrogate(value: string, path: string): void {
  for (let index = 0; index < value.length; index += 1) {
    const unit = value.charCodeAt(index);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      const next = value.charCodeAt(index + 1);
      if (!(next >= 0xdc00 && next <= 0xdfff)) {
        throw new RangeError(`${path} contains an unpaired Unicode surrogate`);
      }
      index += 1;
      continue;
    }
    if (unit >= 0xdc00 && unit <= 0xdfff) {
      throw new RangeError(`${path} contains an unpaired Unicode surrogate`);
    }
  }
}

function normalizeText(value: string, path: string): string {
  const normalized = value.normalize('NFC');
  assertNoUnpairedSurrogate(normalized, path);
  return normalized;
}

function compareUtf16(left: string, right: string): number {
  if (left === right) return 0;
  return left < right ? -1 : 1;
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    return false;
  }
  const prototype = Object.getPrototypeOf(value);
  return prototype === Object.prototype || prototype === null;
}

export function canonicalDecimalToUnits(
  value: string,
  options: { signed?: boolean } = {},
): bigint {
  const signed = options.signed ?? true;
  const raw = normalizeText(String(value ?? '').trim(), 'value');
  const match = /^(-?)(0|[1-9][0-9]*)(?:\.([0-9]+))?$/.exec(raw);
  if (!match) {
    throw new TypeError('value must be canonical decimal text');
  }
  const negative = match[1] === '-';
  if (negative && !signed) {
    throw new RangeError('negative values are not allowed');
  }
  const whole = BigInt(match[2]);
  const fraction = match[3] ?? '';
  const fractionalUnits = BigInt(fraction.slice(0, 6).padEnd(6, '0') || '0');
  const units = whole * KAPPA_SCALE + fractionalUnits;
  return negative && units !== 0n ? -units : units;
}

export function truncDivTowardZero(numerator: bigint, denominator: bigint): bigint {
  if (denominator === 0n) {
    throw new RangeError('division by zero');
  }
  return numerator / denominator;
}

export function multiplyFixed(left: bigint, right: bigint): bigint {
  return truncDivTowardZero(left * right, KAPPA_SCALE);
}

export function divideFixed(left: bigint, right: bigint): bigint {
  return truncDivTowardZero(left * KAPPA_SCALE, right);
}

export function canonicalValue(value: unknown, path = '$'): CanonicalValue {
  if (value === null || typeof value === 'boolean' || typeof value === 'bigint') {
    return value;
  }
  if (typeof value === 'number') {
    throw new TypeError(
      `${path} contains a number; use bigint scaled integers or canonical decimal text`,
    );
  }
  if (typeof value === 'string') {
    return normalizeText(value, path);
  }
  if (Array.isArray(value)) {
    return value.map((item, index) => canonicalValue(item, `${path}[${index}]`));
  }
  if (!isPlainObject(value)) {
    throw new TypeError(`${path} contains an unsupported value`);
  }

  const entries = new Map<string, unknown>();
  for (const [rawKey, item] of Object.entries(value)) {
    const normalizedKey = normalizeText(rawKey, `${path}.<key>`);
    if (entries.has(normalizedKey)) {
      throw new RangeError(`${path} contains duplicate keys after Unicode normalization`);
    }
    entries.set(normalizedKey, item);
  }

  const normalized: CanonicalObject = {};
  for (const key of [...entries.keys()].sort(compareUtf16)) {
    normalized[key] = canonicalValue(entries.get(key), `${path}.${key}`);
  }
  return normalized;
}

function canonicalStringifyNormalized(value: CanonicalValue): string {
  if (value === null) return 'null';
  if (typeof value === 'boolean') return value ? 'true' : 'false';
  if (typeof value === 'bigint') return value.toString(10);
  if (typeof value === 'string') return JSON.stringify(value);
  if (Array.isArray(value)) {
    return `[${value.map(canonicalStringifyNormalized).join(',')}]`;
  }
  const keys = Object.keys(value).sort(compareUtf16);
  return `{${keys
    .map((key) => `${JSON.stringify(key)}:${canonicalStringifyNormalized(value[key])}`)
    .join(',')}}`;
}

export function canonicalBytes(value: unknown): Uint8Array {
  const normalized = canonicalValue(value);
  const bytes = new TextEncoder().encode(canonicalStringifyNormalized(normalized));
  if (bytes.byteLength > MAX_CANONICAL_BYTES) {
    throw new RangeError('canonical payload exceeds the bounded size limit');
  }
  return bytes;
}

export function canonicalSha256(value: unknown): string {
  return createHash('sha256').update(canonicalBytes(value)).digest('hex');
}

export function stateSha256(value: unknown): string {
  const normalized = canonicalValue(value);
  if (!isPlainObject(normalized)) {
    throw new TypeError('state must be a mapping');
  }
  const payload: CanonicalObject = {};
  for (const key of Object.keys(normalized).sort(compareUtf16)) {
    if (!STATE_METADATA_FIELDS.has(key)) {
      payload[key] = normalized[key];
    }
  }
  return canonicalSha256(payload);
}

export function normalizeTransitionTable(value: unknown): TransitionTable {
  if (!isPlainObject(value) || Object.keys(value).length === 0) {
    throw new TypeError('transition_table must be a non-empty mapping');
  }

  const tableEntries = new Map<string, Record<string, unknown>>();
  for (const [rawState, rawActions] of Object.entries(value)) {
    if (!rawState.trim()) {
      throw new TypeError('transition state names must be non-empty strings');
    }
    if (!isPlainObject(rawActions) || Object.keys(rawActions).length === 0) {
      throw new TypeError('each transition state must contain action mappings');
    }
    const state = normalizeText(rawState.trim(), 'transition state');
    if (tableEntries.has(state)) {
      throw new RangeError('duplicate transition state after normalization');
    }
    tableEntries.set(state, rawActions);
  }

  const output: TransitionTable = {};
  for (const state of [...tableEntries.keys()].sort(compareUtf16)) {
    const actionEntries = new Map<string, string>();
    for (const [rawAction, rawTarget] of Object.entries(tableEntries.get(state)!)) {
      if (!rawAction.trim()) {
        throw new TypeError('transition action names must be non-empty strings');
      }
      if (typeof rawTarget !== 'string' || !rawTarget.trim()) {
        throw new TypeError('transition targets must be non-empty strings');
      }
      const action = normalizeText(rawAction.trim(), 'transition action');
      const target = normalizeText(rawTarget.trim(), 'transition target');
      if (actionEntries.has(action)) {
        throw new RangeError('duplicate transition action after normalization');
      }
      actionEntries.set(action, target);
    }
    output[state] = {};
    for (const action of [...actionEntries.keys()].sort(compareUtf16)) {
      output[state][action] = actionEntries.get(action)!;
    }
  }
  return output;
}

function asCanonicalObject(value: unknown, label: string): CanonicalObject {
  const normalized = canonicalValue(value);
  if (!isPlainObject(normalized)) {
    throw new TypeError(`${label} must be a mapping`);
  }
  return normalized;
}

function cloneCanonicalObject(value: CanonicalObject): CanonicalObject {
  return asCanonicalObject(value, 'value');
}

export function transitionPreview(
  currentState: unknown,
  action: unknown,
  transitionTable: unknown,
  options: {
    expectedVersion?: bigint;
    expectedStateHash?: string;
    engineVersion?: string;
  } = {},
): Record<string, unknown> {
  const state = asCanonicalObject(currentState, 'current_state');
  const selectedAction = asCanonicalObject(action, 'action');
  const table = normalizeTransitionTable(transitionTable);

  const currentStatus = state.status;
  const actionType = selectedAction.type;
  if (typeof currentStatus !== 'string' || currentStatus.length === 0) {
    throw new TypeError('current_state.status is required');
  }
  if (typeof actionType !== 'string' || actionType.length === 0) {
    throw new TypeError('action.type is required');
  }

  const currentVersion = state.version ?? 0n;
  if (typeof currentVersion !== 'bigint' || currentVersion < 0n) {
    throw new TypeError('current_state.version must be a non-negative integer');
  }

  const currentHash = stateSha256(state);
  if (
    options.expectedVersion !== undefined &&
    currentVersion !== options.expectedVersion
  ) {
    return {
      ok: false,
      allowed: false,
      status: 'VERSION_CONFLICT',
      currentVersion,
      expectedVersion: options.expectedVersion,
      currentStateHash: currentHash,
      mutationPerformed: false,
    };
  }
  if (
    options.expectedStateHash &&
    currentHash !== options.expectedStateHash
  ) {
    return {
      ok: false,
      allowed: false,
      status: 'STATE_HASH_CONFLICT',
      currentStateHash: currentHash,
      expectedStateHash: options.expectedStateHash,
      mutationPerformed: false,
    };
  }

  const target = table[currentStatus]?.[actionType];
  if (target === undefined) {
    return {
      ok: false,
      allowed: false,
      status: 'TRANSITION_NOT_ALLOWED',
      currentStatus,
      actionType,
      allowedActions: Object.keys(table[currentStatus] ?? {}).sort(compareUtf16),
      currentStateHash: currentHash,
      mutationPerformed: false,
    };
  }

  const requestedTarget = selectedAction.targetStatus;
  if (requestedTarget !== undefined && requestedTarget !== target) {
    return {
      ok: false,
      allowed: false,
      status: 'TARGET_STATUS_MISMATCH',
      currentStatus,
      actionType,
      contractTargetStatus: target,
      requestedTargetStatus: requestedTarget,
      currentStateHash: currentHash,
      mutationPerformed: false,
    };
  }

  const patchValue = selectedAction.patch ?? {};
  if (!isPlainObject(patchValue)) {
    throw new TypeError('action.patch must be a mapping when supplied');
  }
  const patch = asCanonicalObject(patchValue, 'action.patch');
  const protectedFields = new Set([
    'status',
    'version',
    'stateHash',
    'actionHash',
    'chainHash',
    'previousChainHash',
  ]);
  const forbidden = Object.keys(patch)
    .filter((key) => protectedFields.has(key))
    .sort(compareUtf16);
  if (forbidden.length > 0) {
    throw new RangeError(`action.patch contains protected fields: ${forbidden.join(', ')}`);
  }

  const nextState = cloneCanonicalObject(state);
  for (const metadataField of STATE_METADATA_FIELDS) {
    delete nextState[metadataField];
  }
  for (const key of Object.keys(patch).sort(compareUtf16)) {
    nextState[key] = patch[key];
  }
  nextState.status = target;
  nextState.version = currentVersion + 1n;

  const actionHash = canonicalSha256(selectedAction);
  const nextStateHash = stateSha256(nextState);
  const previousChainHash = typeof state.chainHash === 'string' ? state.chainHash : '';
  const engineVersion = normalizeText(
    options.engineVersion?.trim() || 'are-v1',
    'engineVersion',
  );
  const chainMaterial: CanonicalObject = {
    schemaVersion: 'sovereign.are-chain.v1',
    engineVersion,
    sequence: nextState.version,
    previousChainHash,
    actionHash,
    stateHash: nextStateHash,
  };
  const chainHash = canonicalSha256(chainMaterial);

  return {
    ok: true,
    allowed: true,
    status: 'TRANSITION_VALIDATED',
    currentStatus,
    nextStatus: target,
    currentVersion,
    nextVersion: nextState.version,
    currentStateHash: currentHash,
    actionHash,
    nextStateHash,
    previousChainHash,
    chainHash,
    stateHashContract: 'canonical-state-without-chain-metadata',
    nextState,
    mutationPerformed: false,
    truthNotice:
      'Pure transition preview only; persistence, effects and runtime success are not claimed.',
  };
}

export function replayVerify(
  initialState: unknown,
  actions: unknown,
  transitionTable: unknown,
  options: {
    expectedFinalStateHash?: string;
    engineVersion?: string;
  } = {},
): Record<string, unknown> {
  if (!Array.isArray(actions)) {
    throw new TypeError('actions must be a list');
  }
  if (actions.length > MAX_REPLAY_ACTIONS) {
    throw new RangeError(`actions exceed the bounded limit of ${MAX_REPLAY_ACTIONS}`);
  }

  let state = asCanonicalObject(initialState, 'initial_state');
  const steps: Record<string, unknown>[] = [];
  for (let index = 0; index < actions.length; index += 1) {
    const version = state.version ?? 0n;
    if (typeof version !== 'bigint') {
      throw new TypeError('state version must be bigint');
    }
    const result = transitionPreview(state, actions[index], transitionTable, {
      expectedVersion: version,
      expectedStateHash: stateSha256(state),
      engineVersion: options.engineVersion,
    });
    steps.push({
      index: BigInt(index + 1),
      allowed: Boolean(result.allowed),
      status: result.status,
      actionHash: result.actionHash,
      stateHash: result.nextStateHash,
      chainHash: result.chainHash,
    });
    if (!result.allowed) {
      return {
        ok: false,
        status: 'REPLAY_BLOCKED',
        failedStep: BigInt(index + 1),
        steps,
        finalState: state,
        finalStateHash: stateSha256(state),
        finalChainHash: typeof state.chainHash === 'string' ? state.chainHash : null,
        stateHashContract: 'canonical-state-without-chain-metadata',
        mutationPerformed: false,
        crossRuntimeParityProven: false,
      };
    }
    state = cloneCanonicalObject(result.nextState as CanonicalObject);
    state.chainHash = result.chainHash as string;
  }

  const finalStateHash = stateSha256(state);
  const matches =
    !options.expectedFinalStateHash ||
    finalStateHash === options.expectedFinalStateHash;
  return {
    ok: matches,
    status: matches ? 'REPLAY_VERIFIED' : 'FINAL_STATE_HASH_MISMATCH',
    steps,
    finalState: state,
    finalStateHash,
    finalChainHash: typeof state.chainHash === 'string' ? state.chainHash : null,
    stateHashContract: 'canonical-state-without-chain-metadata',
    expectedFinalStateHash: options.expectedFinalStateHash || null,
    mutationPerformed: false,
    crossRuntimeParityProven: false,
    truthNotice:
      'Independent TypeScript reference replay; parity is established only by comparing the same committed vectors with the Python reference.',
  };
}

export function materializeTaggedIntegers(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map(materializeTaggedIntegers);
  }
  if (!isPlainObject(value)) {
    return value;
  }
  const keys = Object.keys(value);
  if (
    keys.length === 1 &&
    keys[0] === '$integer' &&
    typeof value.$integer === 'string'
  ) {
    if (!/^-?(?:0|[1-9][0-9]*)$/.test(value.$integer)) {
      throw new TypeError('tagged integer must contain canonical decimal integer text');
    }
    return BigInt(value.$integer);
  }
  const output: Record<string, unknown> = {};
  for (const key of keys) {
    output[key] = materializeTaggedIntegers(value[key]);
  }
  return output;
}

export function tagBigInts(value: unknown): unknown {
  if (typeof value === 'bigint') {
    return { $integer: value.toString(10) };
  }
  if (Array.isArray(value)) {
    return value.map(tagBigInts);
  }
  if (isPlainObject(value)) {
    const output: Record<string, unknown> = {};
    for (const key of Object.keys(value).sort(compareUtf16)) {
      output[key] = tagBigInts(value[key]);
    }
    return output;
  }
  return value;
}
